core/src/dird/ua_acl.cc

```diff
@@ -47,132 +47,101 @@ bool UaContext::AclAccessOk(int acl, const char* item, bool audit_event)
 {
   return std::string::npos != string_to_check.find_first_of(".()[]|^$+?*");
 }
 
+/**
+ * acl_list_value: value of the ACL definition.
+ * acl_list_compare_value: value ot compare against. This is identical to
+ * acl_list_value or a modified acl_list_value to compare against.
+ */
+static inline bool CompareAclListValueWithItem(
+    int acl,
+    const char* acl_list_value,
+    const char* acl_list_compare_value,
+    const char* item)
+{
+  // gives full access
+  if (Bstrcasecmp("*all*", acl_list_compare_value)) {
+    Dmsg2(1400, "Global ACL found in %d %s\n", acl, acl_list_value);
+    return true;
+  }
+
+  if (Bstrcasecmp(item, acl_list_compare_value)) {
+    // Explicit match.
+    Dmsg3(1400, "ACL found %s in %d %s\n", item, acl, acl_list_value);
+    return true;
+  }
+
+  /* If we didn't get an exact match see if we can use the pattern as a
+   * regex. */
+  if (is_regex(acl_list_compare_value)) {
+    regex_t preg{};
+    int match_length = strlen(item);
+    int rc = regcomp(&preg, acl_list_compare_value, REG_EXTENDED | REG_ICASE);
+    if (rc != 0) {
+      // Not a valid regular expression so skip it.
+      Dmsg1(1400, "Not a valid regex %s, ignoring for regex compare\n",
+            acl_list_value);
+      return false;
+    }
+
+    int nmatch = 1;
+    regmatch_t pmatch[1]{};
+    if (regexec(&preg, item, nmatch, pmatch, 0) == 0) {
+      // Make sure its not a partial match but a full match.
+      Dmsg2(1400, "Found match start offset %d end offset %d\n",
+            pmatch[0].rm_so, pmatch[0].rm_eo);
+      if ((pmatch[0].rm_eo - pmatch[0].rm_so) >= match_length) {
+        Dmsg3(1400, "ACL found %s in %d using regex %s\n", item, acl,
+              acl_list_value);
+        regfree(&preg);
+        return true;
+      }
+    }
+    regfree(&preg);
+  }
+  return false;
+}
+
 /**
  * Loop over the items in the alist and verify if they match the given item
  * that access was requested for.
  */
-static inline bool FindInAclList(alist<const char*>* list,
-                                 int acl,
-                                 const char* item,
-                                 int len)
+static inline std::optional<bool> FindInAclList(alist<const char*>* list,
+                                                int acl,
+                                                const char* item,
+                                                int len)
 {
-  int rc;
-  regex_t preg{};
-  int nmatch = 1;
-  bool retval = false;
-  regmatch_t pmatch[1]{};
-  const char* list_value;
-
   // See if we have an empty list.
   if (!list) {
-    /*
-     * Empty list for Where => empty where accept anything.
-     * For any other list, reject everything.
-     */
+    /* Empty list for Where => empty where accept anything.
+     * For any other list, reject everything. */
     if (len == 0 && acl == Where_ACL) {
       Dmsg0(1400, "Empty Where_ACL allowing restore anywhere\n");
-      retval = true;
+      return true;
```

Comment:
You're absolutely right.
Currently an empty WhereACL have not been handled, as the code for this has been wrong.
So I decided to remove this special treatment at all, as, even if implemented correctly would cause problems if used together with profiles (as emptiness would be check for each resource separately).